### ISSUE TITLE
Subsystem benchmarks: determine node CPU usage for 1000 validators and 200 full occupied cores

### ISSUE THREAD

**Issue by sandreim** [paritytech/polkadot-sdk#5035]
We should repeat the testing done in https://github.com/paritytech/polkadot-sdk/issues/4126#issuecomment-2084452189 to get some numbers with the 1kv and 200 cores. This woulld help to better inform the decision and outlook of raising HW specs: https://forum.polkadot.network/t/rfc-increasing-recommended-minimum-core-count-for-reference-hardware/8156

**Comment by sandreim** [comment]
Thanks @alexggh! More than 60% of a nodes CPU used for PVF execution sounds amazing.

You assumed 1 full CPU being used by networking, however with` libp2p`, it's usage was 250% in versi tests. But we don't know any numbers for `litep2p`. I think we need to do some benchmarking for the substrate network with similar load as we've done in this test.

**Comment by burdges** [comment]
> approval-distribution                           2.1300

Is this just because it's lots of messages? 

> approval-voting                                    3.0523

Is this verifying assignment VRFs, verifying approval votes, and managing the approvals database?

> availability-distribution                         0.0347

Is this just sending out the chunks?

> availability-store                                    0.1906

What is here?

> bitfield-distribution                               0.0484
> availability-recovery                              1.4412

All erasure code work lies here, no?

We've already using some flavor of systemic chunks, or at least direct fetch here, so this mostly counts only recomputing the availability merkle root.  If many nodes go offline then we'd be doing reconstructions here, which doubles or tripples this, right?

> statement-distribution                           0.0826

This is all in backing, right?

**Comment by alexggh** [comment]
> > approval-distribution 2.1300

> Is this just because it's lots of messages?

Yes, bookkeeping and the logic to gossip scales up with the number of messages and the number of validators in the network.

> > approval-voting 3.0523

> Is this verifying assignment VRFs, verifying approval votes, and managing the approvals database?

Yes.

> > availability-distribution 0.0347

> Is this just sending out the chunks? 

Yes. 

> > availability-store 0.1906

> What is here?

Storing the chunks


> > bitfield-distribution 0.0484
> > availability-recovery 1.4412

> All erasure code work lies here, no?

Yes

> We've already using some flavor of systemic chunks, or at least direct fetch here, so this mostly counts only recomputing the availability merkle root. If many nodes go offline then we'd be doing reconstructions here, which doubles or tripples this, right?

I have to double check, but I don't think the system-chunks are enabled in this benchmarks.

> > statement-distribution 0.0826

> This is all in backing, right?

Yes.

**Comment by AndreiEres** [comment]
> > We've already using some flavor of systemic chunks, or at least direct fetch here, so this mostly counts only recomputing the availability merkle root. If many nodes go offline then we'd be doing reconstructions here, which doubles or tripples this, right?
> 
> I have to double check, but I don't think the system-chunks are enabled in this benchmarks.

In test used `BackersFirstAlways` recovery strategy. I assume there is a little chance of chunks because of deterministic nature of benchmarks.

**Comment by alexggh** [comment]
Related to this I've been running a small versi experiment.

<h2>Running polkadot in a cpu restrictive environment</h2>
<p>In order to determine the stability, readiness and requirements of polkadot for 1k validators and 200 cores, I ran a few experiments where the node hardware requirements where scaled down by a factor of 16 and 8 and pushed the network to the limit.</p>
<h2>Translating dimensions by the scale factor</h2>

Dimension | Real | Scaled down by 16 | Scaled down by 8
-- | -- | -- | --
CPU allocation | 8 cores | 0.5 cores | 1 core
Number of validators | 1000 | 64 | 128
Number of full cores/parachains | 200 | 13 | 25
Pov Execution time | ~2s | ~2s | ~2s
PoV size | ~4MiB | ~2 MiB | ~2 MiB


<p>The goals here was that by scaling every dimension we would gather useful data about the behaviour of the real network, because validators would be CPU starved/throttled, so everything would take longer to execute be delayed.</p>
<p>!!Limitations!! This would probably not reflect very well the load created by things that are scaling with the number of validators and parachains, like assignments and approvals or bitfield processing.</p>
<h2>Results</h2>
<p>I ended up running three experiments:</p>
<ol>
<li>Everything scaled down by 16.</li>
<li>Everything scaled down by 8</li>
<li>Everything scaled down by 8, with litep2p</li>
</ol>
<h3>Findings / Confirming what we already assumed or knew</h3>
<ul>
<li>During the whole test the network was stable, block production and finality were stable.</li>
<li>The biggest CPUs hogs by an order of magnitude are <code>libp2p-node</code> and <code>erasure-node</code> tasks.</li>
<li>When scaled by a factor of 16(0.5cpu), only 8 parachains could produce blocks at around ~6s, after that adding more parachains, slowed block production for parachains because availability started being slower, so with the target of 13 parachains, parachain blocks were being produced at around 9s rate.
<ul>
<li>Increasing the CPU allocation from 0.5 to 1, automatically made parachains producing blocks at 6s rate</li>
</ul>
</li>
<li>Adding more validators to a network, it is automatically reducing the number of parachain blocks a validators has to check for approving.</li>
<li>When scaled by a factor of 8(1cpu) we were able to support 25 parachains parachains producing blocks at 6s rate.
<ul>
<li>Total Time to recover PoV was around 1s and 5s, that means that in this case when we have all parachains producing really big PoVs we were spending more time on PoV recovery than on PoV execution, that raises the question if we need to back-pressure in case nodes start to fall behind on PoV recovery, although I think this is just an artefact of this synthetic scenario.</li>
<li>Switching the same network to the litep2p, made parachains slow down from 6s rate to around 11s, availability seems to slow down because of this <a href="https://grafana.teleport.parity.io/goto/jqUOYteSR?orgId=1">ChannelClogged</a>, to be investigated if this is an artefact of the synthetic scenario or something.</li>
</ul>
</li>
</ul>

**Comment by sandreim** [comment]
Nice work Alex!

Based on these numbers we should be fine with 1kV and 200 cores, but that also depends on how we deal with the high amount of gossip messages and chunk requests. Subsystem benchmarking should provide a good estimate for this.

What are you planning to do next ? 

> * Switching the same network to the litep2p, made parachains slow down from 6s rate to around 11s, availability seems to slow down because of this [ChannelClogged](https://grafana.teleport.parity.io/goto/jqUOYteSR?orgId=1), to be investigated if this is an artefact of the synthetic scenario or something.

Maybe libp2p might be using some unbounded or higher limit channels, litep2p has them [bounded at 4096 ](https://github.com/paritytech/litep2p/blob/a68b7136b6579766508d171371be1b3b3d60751f/src/lib.rs#L94)

CC @paritytech/networking

**Comment by alexggh** [comment]
> Based on these numbers we should be fine with 1kV and 200 cores, but that also depends on how we deal with the high amount of gossip messages and chunk requests. Subsystem benchmarking should provide a good estimate for this.
> What are you planning to do next ?

That's already estimated here: https://github.com/paritytech/polkadot-sdk/issues/5035#issuecomment-2232988942 and we concluded with 8 cores we should  have enough CPUs for it.

So once all the optimisations that we have in progress land on kusama, we should be able to gradually ramp up the number of validators doing parachains consensus work there and do it gradually so we can catch any deterioration and reverse curse  if our estimations prove wrong.

**Comment by burdges** [comment]
> The biggest CPUs hogs by an order of magnitude are libp2p-node and erasure-node tasks.

An order of magnitude?  wow.  I suppose erasure-node acts like a fixed cost here, due to how backers work.  And libp2p-node might've many fixed costs too.  That'd be roughly 16x or 8x.

In producton, what hogs the resources? Someone claimed signature verification recently.

**Comment by alexggh** [comment]
> In producton, what hogs the resources? Someone claimed signature verification recently.

These are the usages of a real kusama validators.
<img width="464" alt="Screenshot 2024-09-11 at 10 10 32" src="https://github.com/user-attachments/assets/71e7693f-3a98-4875-b0bf-67eb05b7f290">

And these are the usages that I got in restrictive environment:
<img width="453" alt="Screenshot 2024-09-11 at 10 13 47" src="https://github.com/user-attachments/assets/cd2f23b7-4b75-4eba-85bb-a26be4614096">

The PoVs on Kusama are really small when compared with my tests hence why I think erasure-task does not show up as using a lot of cpu on kusama.

> Someone claimed signature verification recently.

You mean assignments and approval, yeah that's a problem as well especially if you have a lot of no-shows, we did a lot of optimisations on that path as part of this thread: https://github.com/paritytech/polkadot-sdk/pull/4849, so it shouldn't be a problem anymore.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/litep2p-network-backend-updates/9973/1
